std/assembly/atomics.ts

```diff
@@ -0,0 +1,76 @@
+export namespace Atomics {
+  // @ts-ignore: decorator
+  @inline
+  export function load<T>(array: ArrayBufferView<T>, index: i32): T {
+    return atomic.load<T>(changetype<usize>(array.buffer) + (index << alignof<T>()) + array.byteOffset);
+  }
+
+  // @ts-ignore: decorator
+  @inline
+  export function store<T>(array: ArrayBufferView<T>, index: i32, value: T): void {
+    atomic.store<T>(changetype<usize>(array.buffer) + (index << alignof<T>()) + array.byteOffset, value, index);
+  }
+
+  // @ts-ignore: decorator
+  @inline
+  export function add<T>(array: ArrayBufferView<T>, index: i32, value: T): T {
+    return atomic.add<T>(changetype<usize>(array.buffer) + (index << alignof<T>()) + array.byteOffset, value, index);
+  }
+
+  // @ts-ignore: decorator
+  @inline
+  export function sub<T>(array: ArrayBufferView<T>, index: i32, value: T): T {
+    return atomic.sub<T>(changetype<usize>(array.buffer) + (index << alignof<T>()) + array.byteOffset, value, index);
+  }
+
+  // @ts-ignore: decorator
+  @inline
+  export function and<T>(array: ArrayBufferView<T>, index: i32, value: T): T {
+    return atomic.and<T>(changetype<usize>(array.buffer) + (index << alignof<T>()) + array.byteOffset, value, index);
+  }
+
+  // @ts-ignore: decorator
+  @inline
+  export function or<T>(array: ArrayBufferView<T>, index: i32, value: T): T {
+    return atomic.or<T>(changetype<usize>(array.buffer) + (index << alignof<T>()) + array.byteOffset, value, index);
+  }
+
+  // @ts-ignore: decorator
+  @inline
+  export function xor<T>(array: ArrayBufferView<T>, index: i32, value: T): T {
+    return atomic.xor<T>(changetype<usize>(array.buffer) + (index << alignof<T>()) + array.byteOffset, value, index);
+  }
+
+  // @ts-ignore: decorator
+  @inline
+  export function exchange<T>(array: ArrayBufferView<T>, index: i32, value: T): T {
+    return atomic.xchg<T>(changetype<usize>(array.buffer) + (index << alignof<T>()) + array.byteOffset, value, index);
+  }
+
+  // @ts-ignore: decorator
+  @inline
+  export function compareExchange<T>(array: ArrayBufferView<T>, index: i32, expectedValue: T, replacementValue: T): T {
+    return atomic.cmpxchg<T>(
+      changetype<usize>(array.buffer) + (index << alignof<T>()) + array.byteOffset,
+      expectedValue,
+      replacementValue,
+      index
+    );
+  }
+
+  // @ts-ignore: decorator
+  @inline
+  export function wait<T>(array: ArrayBufferView<T>, value: T, timeout: i64 = -1): AtomicWaitResult {
```

Comment:
There should be a compile time check here.

```ts
if (sizeof<T>() < 4) ERROR("Can only call Atomics.wait on 32 or 64 bit pointers.");
if (!isInteger<T>()) ERROR("Must be called on an integer type.");
```

Otherwise this will result in a wasm binary that doesn't validate.

Reply:
Obviously, please change the error descriptions. :)

Reply:
Wondering if this doesn't already error when trying to compile `atomic.wait<T>` in the function body?

Reply:
I remember hitting some kind of validation error. I am probably mistaken as @dcodeIO suggests because it was probably related to the `--enable threads` flag and combining that with the experimental node flag. Forgive me. 😅

Reply:
@jtenner 
> if (sizeof<T>() < 4) ERROR("Can only call Atomics.wait on 32 or 64 bit pointers.");

Why? By spec typedArray in Atomics could be any TypedArray except Float32Array and Float64Array